File: Loenn/entities/distortionField.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local distortionField = {}

distortionField.name = "MaggyHelper/DistortionField"
distortionField.fieldInformation = {
    distortionType = {
        options = {"Reverse", "Random", "GravityFlip", "SlowMotion"},
        editable = false
    },
    intensity = {
        minimumValue = 0,
        maximumValue = 1
    }
}
distortionField.placements = {
    {
        name = "reverse",
        data = {
            distortionType = "Reverse",
            intensity = 1.0,
            width = 32,
            height = 32
        }
    },
    {
        name = "random",
        data = {
            distortionType = "Random",
            intensity = 1.0,
            width = 32,
            height = 32
        }
    },
    {
        name = "gravity_flip",
        data = {
            distortionType = "GravityFlip",
            intensity = 1.0,
            width = 32,
            height = 32
        }
    },
    {
        name = "slow_motion",
        data = {
            distortionType = "SlowMotion",
            intensity = 0.5,
            width = 32,
            height = 32
        }
    }
}

function distortionField.sprite(room, entity)
    local sprites = {}
    local width = entity.width or 32
    local height = entity.height or 32
    
    for x = 0, width - 8, 8 do
        for y = 0, height - 8, 8 do
            local sprite = drawableSprite.fromTexture("effects/distortion_field/default", entity)
            sprite:setPosition(entity.x + x, entity.y + y)
            sprite:setJustification(0, 0)
            sprite.color = {0.5, 0.0, 0.5, 0.3}
            table.insert(sprites, sprite)
        end
    end
    
    return sprites
end

function distortionField.selection(room, entity)
    return utils.rectangle(entity.x, entity.y, entity.width or 32, entity.height or 32)
end

return distortionField

File: framework/extracted/Ahorn/entities/DistortionField.jl
module DistortionField

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/DistortionField" DistortionField(x::Integer, y::Integer, distortionType::String="Reverse", height::Integer=32, intensity::Number=1.0, width::Integer=32)

const placements = Ahorn.PlacementDict(
    "reverse" => Ahorn.EntityPlacement(DistortionField),
    "random" => Ahorn.EntityPlacement(DistortionField),
    "gravity_flip" => Ahorn.EntityPlacement(DistortionField),
    "slow_motion" => Ahorn.EntityPlacement(DistortionField)
)

function Ahorn.selection(entity::DistortionField)
    return Ahorn.getEntityRectangle(entity)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::DistortionField, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.getEntityRectangle(entity), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end
